Move MountsInGlobalPDPath from mount pkg to volume

codenrhoden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Since `pkg/util/mount` is going to move out of k/k, this exported constant
that is Kubernetes specific needs to move somewhere else. Made sense to
move it to `pkg/volume/util`.

For the volume plugins that call `mounter.GetDeviceNameFromMount()`,
rather than changing the signature of `mounter.GetDeviceNameFromMount()`
just have the volume plugins append `MountsInGlobalPDPath` to the volume
pluginDir themselves rather than having `mounter.GetDeviceNameFromMount()`
do it since that code no longer has access to that constant.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
This is another step in making #68513 be a lot simpler.

This PR does not attempt to address #74548 at all, just maintain current functionality.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

codenrhoden commented:
/sig storage
/assign @msau42 @saad-ali

msau42 commented on pkg/util/mount/mount_windows.go:
```diff
@@ -209,10 +209,9 @@ func getDeviceNameFromMount(mounter Interface, mountPath, pluginDir string) (str
 	for _, ref := range refs {
-		if strings.Contains(ref, basemountPath) {
-			volumeID, err := filepath.Rel(normalizeWindowsPath(basemountPath), ref)
+		if strings.Contains(ref, pluginDir) {
```
does plugindir need to be normalized first?

codenrhoden replied:
Yes, it does. Good catch.

msau42 commented on pkg/volume/awsebs/aws_ebs.go:
```diff
@@ -250,7 +250,7 @@ func getVolumeSource(
 func (plugin *awsElasticBlockStorePlugin) ConstructVolumeSpec(volName, mountPath string) (*volume.Spec, error) {
 	mounter := plugin.host.GetMounter(plugin.GetPluginName())
-	pluginDir := plugin.host.GetPluginDir(plugin.GetPluginName())
+	pluginDir := filepath.Join(plugin.host.GetPluginDir(plugin.GetPluginName()), util.MountsInGlobalPDPath)
```
does it seem like enough plugins use this to warrant some new "GetPluginMountDir" interface?

codenrhoden replied:
I'm not sure it would need to be an interface, rather just an exported function perhaps in `pkg/volume/util`?

```go
func GetPluginMountDir(host volume.VolumeHost, name string) string {
    pluginDir := filepath.Join(host.GetPluginDir(name), MountsInGlobalPDPath)
    return pluginDir
}
```

Something like that?

msau42 replied:
This sounds reasonable to me

bertinatto commented on pkg/util/mount/mount_linux.go:
```diff
@@ -318,10 +318,9 @@ func getDeviceNameFromMount(mounter Interface, mountPath, pluginDir string) (str
-		if strings.HasPrefix(ref, basemountPath) {
-			volumeID, err := filepath.Rel(basemountPath, ref)
+		if strings.HasPrefix(ref, pluginDir) {
+			volumeID, err := filepath.Rel(pluginDir, ref)
```
Seems like `pluginDir`, that comes from the plugin's `ConstructVolumeSpec`, is no longer only the plugin directory. Now it has the `MountsInGlobalPDPath` const appended to it, right?

We might have to change the `mount.Interface` as well.

codenrhoden replied:
That was what I was going for here, yes. I was trying to make as few changes as possible, without having to change `mount.Interface`.

An alternative would be to go ahead and change the interface from:
```go
GetDeviceNameFromMount(mountPath, pluginDir string)
```
to
```go
GetDeviceNameFromMount(mountPath, pluginDir, mountDir string)
```
And then append the mount dir in the function, which is what it did before.

The overall goal here is not have anything in `pkg/util/mount` depend on access to that constant, which needs to live in another package (I had put it in `pkg/volume/util`).

I'm fine with either method.

msau42 replied:
I prefer changing it to the first

msau42 replied:
Sorry I wasn't very clear.  I preferred only having two arguments to `GetDeviceNameFromMount`, and having some helper method that appends the constant to the plugin dir.

codenrhoden replied:
doh. okay, I can do that. 😆

codenrhoden commented on pkg/volume/azure_dd/azure_common.go:
```diff
@@ -80,7 +80,7 @@ func makeGlobalPDPath(host volume.VolumeHost, diskUri string, isManaged bool) (s
 	// "{m for managed b for blob}{hashed diskUri or DiskId depending on disk kind }"
 	diskName := fmt.Sprintf(uniqueDiskNameTemplate, prefix, hashedDiskUri)
-	pdPath := filepath.Join(host.GetPluginDir(azureDataDiskPluginName), mount.MountsInGlobalPDPath, diskName)
+	pdPath := filepath.Join(host.GetPluginDir(azureDataDiskPluginName), util.MountsInGlobalPDPath, diskName)
```
Probably. Though if I an make a single exported function to do it, as I suggested in response to @msau42 comment, there will be only one version.

msau42 replied:
I think it's actually important for windows to use filepath instead of path.  @andyzhangx can you clarify the difference?

andyzhangx replied:
Pls always use `filepath.Join` since `path.Join` may not work on Windows.

codenrhoden commented:
@msau42 @bertinatto 

Thanks for the reviews, and apologies for the slow response. Both of my kids have been sick this week, so I've been unable to come back to this.

If I have any chance of getting this in before code freeze, I'd like to try for that.

I think the main question is whether or not to keep `mount.Interface` the way it is now, which requires callers of `mounter.GetDeviceNameFromMount()` to append the mountDir constant themselves before the call (which could be done with a common function), or to change the interface to take the mountDir as a third parameter.

I don't have a preference one way or the other. I had leaned the way I did because I didn't want to change the interface if I could avoid it in this PR.

andyzhangx commented on pkg/volume/photon_pd/photon_pd.go:
```diff
@@ -285,7 +286,7 @@ func (c *photonPersistentDiskUnmounter) TearDownAt(dir string) error {
 
 func makeGlobalPDPath(host volume.VolumeHost, devName string) string {
-	return path.Join(host.GetPluginDir(photonPersistentDiskPluginName), mount.MountsInGlobalPDPath, devName)
+	return path.Join(host.GetPluginDir(photonPersistentDiskPluginName), util.MountsInGlobalPDPath, devName)
```
filepath.Join

andyzhangx commented on pkg/volume/storageos/storageos.go:
```diff
@@ -441,7 +441,7 @@ func (b *storageosMounter) SetUpAt(dir string, fsGroup *int64) error {
 
 func makeGlobalPDName(host volume.VolumeHost, pvName, volNamespace, volName string) string {
-	return path.Join(host.GetPluginDir(utilstrings.EscapeQualifiedName(storageosPluginName)), mount.MountsInGlobalPDPath, pvName+"."+volNamespace+"."+volName)
+	return path.Join(host.GetPluginDir(utilstrings.EscapeQualifiedName(storageosPluginName)), util.MountsInGlobalPDPath, pvName+"."+volNamespace+"."+volName)
```
filepath.Join

andyzhangx commented on pkg/volume/rbd/rbd_util.go:
```diff
@@ -255,7 +256,7 @@ func makePDNameInternal(host volume.VolumeHost, pool string, image string) strin
 	}
 	// Return the canonical format path.
-	return path.Join(host.GetPluginDir(rbdPluginName), mount.MountsInGlobalPDPath, pool+"-image-"+image)
+	return path.Join(host.GetPluginDir(rbdPluginName), volutil.MountsInGlobalPDPath, pool+"-image-"+image)
```
filepath.Join

codenrhoden committed:
Move MountsInGlobalPDPath from mount pkg to volume

Since pkg/util/mount is going to move out of k/k, this exported constant
that is Kubernetes specific needed to move somewhere else. Made sense to
move it to pkg/volume/util.

Update GetDeviceNameFromMount in the mount interface to now take a
pluginMountDir argument, which is volume plugin dir with the global
mount path appended to it already.

codenrhoden committed:
Always use filepath.Join instead of path.Join

This patch cleans up pkg/util/mount/* and pkg/util/volume/* to always
use filepath.Join instead of path.Join. filepath.Join is preferred
because path.Join can have issues on Windows.

codenrhoden commented:
@msau42 Okay, I went ahead and did this again after changing the interface. That commit is probably a bit easier to review now.

I ran with @andyzhangx's guidance to always use `filepath.Join`, and went ahead and cleaned up all usage in `pkg/util/mount` and `pkg/util/volume` in a second/separate commit.

codenrhoden commented:
/test pull-kubernetes-e2e-gce

codenrhoden commented:
/test pull-kubernetes-bazel-test

msau42 commented:
/lgtm

@andyzhangx @jingxu97 would you be able to help test this change on windows? Especially the configmap/secret/emptydir/projected/atomic writer changes?

codenrhoden commented:
/retest

codenrhoden commented:
Okay, this is all good to go after a needed rebase.

@andyzhangx @jingxu97 any chance you'll be able to take a look at this one?

codenrhoden commented:
/retest

dims commented:
/milestone v1.15

dims commented:
/priority important-soon

codenrhoden commented:
I have no idea why, after every time I rebase and force-push new commits, that the tests immediately fail because of being unable to merge.  For example:

```
I0417 19:27:05.980] Call: git merge --no-ff -m 'Merge +refs/pull/74734/head:refs/pr/74734' 067c79747793b879f93aa74766d83863f1e1aaf3
W0417 19:27:06.009] merge: 067c79747793b879f93aa74766d83863f1e1aaf3 - not something we can merge
```

Testing this locally says it can merge just fine. And then doing `/retest` it merges and tests fine. Super weird.

codenrhoden commented:
/retest

codenrhoden commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

andyzhangx commented:
/test pull-kubernetes-integration

andyzhangx commented on pkg/volume/awsebs/aws_ebs.go:
```diff
@@ -250,8 +250,8 @@ func getVolumeSource(
 
 func (plugin *awsElasticBlockStorePlugin) ConstructVolumeSpec(volName, mountPath string) (*volume.Spec, error) {
 	mounter := plugin.host.GetMounter(plugin.GetPluginName())
-	pluginDir := plugin.host.GetPluginDir(plugin.GetPluginName())
```
is the old `plugin.host.GetPluginDir` func removed now? If not, I would suggest keep the original code, make this PR as small as possible, easy for review.

codenrhoden replied:
I may move the `path.Join` to `filepath.Join` code to a separate PR. I don't know what happened last night, but my late-night rebase squashed my two commits into one (and I know I didn't do that on purpose).

With that stuff taken out, it's pretty small.

To your specific question, no `GetPluginDir` is not removed. In fact is called from within the new `makeGlobalPDPath()`.

andyzhangx commented:
/test pull-kubernetes-cross

codenrhoden commented on pkg/util/mount/mount_windows.go:
```diff
@@ -212,7 +212,7 @@ func getDeviceNameFromMount(mounter Interface, mountPath, pluginDir string) (str
 	if len(refs) == 0 {
 		return "", fmt.Errorf("directory %s is not mounted", mountPath)
 	}
-	basemountPath := normalizeWindowsPath(path.Join(pluginDir, MountsInGlobalPDPath))
```
yes, callers within K8s have to append `MountsInGlobalPDPath` before making this call now. Picture this code living at `k8s.io/utils/mount` instead of within `k8s.io/kubernetes`. It can't/won't know anything about `MountsInGlobalPDPath`.

I made sure to update all callers.

codenrhoden commented:
Hi @andyzhangx,

I needed to rebase again due to merge conflicts, so I went ahead and corrected whatever I messed up last night and this PR is back into two commits. That should make it much easier to review. For convenience, the commit with just the code related to moving the constant is [here](https://github.com/kubernetes/kubernetes/pull/74734/commits/674508729d882466c63f5e91095f27e9354a3fbc).

Can you take a look again?

The code has not changed since you last looked, but hopefully it is easier to see the full picture of what is changing now. The approach to have all callers to `mount.GetDeviceNameFromMount` append `MountsInGlobalPDPath` themselves (via a new helper function) was discussed with @msau42 previously in this PR as the preferred path forward.

Thanks!

codenrhoden commented:
/test pull-kubernetes-e2e-gce-csi-serial

andyzhangx commented:
/test pull-kubernetes-cross

andyzhangx reviewed (approved):
/lgtm

jingxu97 commented:
/test pull-kubernetes-e2e-windows-gce

codenrhoden commented:
/test pull-kubernetes-e2e-windows-gce

codenrhoden commented:
Thanks @msau42, you beat me to it. Whatever the test is, the test is broken. =)

yujuhong commented:
> Thanks @msau42, you beat me to it. Whatever the test is, the test is broken. =)

The test requires a new kubetest image to pick up the change adding the flag. I filed an issue in test-infra to build and push a new image: https://github.com/kubernetes/test-infra/issues/12340

jsafrane commented:
/approve

codenrhoden commented:
/test pull-kubernetes-bazel-test

yujuhong commented:
/test pull-kubernetes-e2e-windows-gce

codenrhoden commented:
/test pull-kubernetes-e2e-windows-gce

yujuhong commented:
/test pull-kubernetes-e2e-windows-gce

We are trying to get the test job working, so I'm hijacking this PR to run some tests

yujuhong commented:
/test pull-kubernetes-e2e-windows-gce

codenrhoden commented:
@yujuhong Hijack away, and good luck. =)

codenrhoden commented:
I doubt it is helpful, but the main thing that jumps out to me from the logs to me is the linux reference, here:

`--config=cross:linux_amd64`

For a windows test, that raises red flags for me.

yujuhong commented:
/test pull-kubernetes-e2e-windows-gce

yujuhong commented:
> For a windows test, that raises red flags for me.

We need to build both windows and linux artifacts since the master control plane runs only on linux :-)

codenrhoden commented:
Anything I can do to move this along?

msau42 commented:
windows pull job still needs a bit of work to get passing
/skip

msau42 commented:
/lgtm

yujuhong commented:
> windows pull job still needs a bit of work to get passing

It's not required, so you can safely ignore it.

Our build/stage process have poor support of handling multiple platforms simultaneously. I made some progress for building them, but more changes are needed for staging...
